presto-main/src/main/java/com/facebook/presto/execution/scheduler/ConsistentHashingNodeProvider.java

```diff
@@ -56,14 +58,21 @@ public ConsistentHashingNodeProvider(NavigableMap<Integer, InternalNode> candida
     public List<HostAddress> get(String key, int count)
     {
         ImmutableList.Builder<HostAddress> nodes = ImmutableList.builder();
-        for (int i = 0; i < count; i++) {
-            int hashKey = HASH_FUNCTION.hashString(format("%s%d", key, i), UTF_8).asInt();
+        Set<HostAddress> unique = new HashSet<>();
+        int i = 0;
+        while (unique.size() < count) {
+            int hashKey = HASH_FUNCTION.hashString(format("%s%d", key, i++), UTF_8).asInt();
             Map.Entry<Integer, InternalNode> entry = candidates.ceilingEntry(hashKey);
+            HostAddress candidate;
             if (entry != null) {
-                nodes.add(candidates.ceilingEntry(hashKey).getValue().getHostAndPort());
+                candidate = candidates.ceilingEntry(hashKey).getValue().getHostAndPort();
             }
             else {
-                nodes.add(candidates.firstEntry().getValue().getHostAndPort());
+                candidate = candidates.firstEntry().getValue().getHostAndPort();
+            }
+            if (!unique.contains(candidate)) {
```

Comment:
Instead of building the node list here, why not add to the unique set no matter what, and then convert the set to an immutable list before returning it?

Reply:
ordering is important here.